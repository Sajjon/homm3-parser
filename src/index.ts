import { parse, StreamInput, TagOrWrapper } from 'binary-markup'
import * as pako from 'pako'
import { h3mFile } from './parsers/h3m/h3m.bml'
import { H3MFile } from './parsers/h3m/contracts'

export * from './parsers/assets'
export * from './parsers/h3m/contracts'
export * from './parsers/h3m/constants'

export function parseH3M(data: StreamInput): H3MFile {
  return parse<H3MFile>(h3mFile, data)
}

export function parseH3MFile(data: Uint8Array): H3MFile {
  const content = pako.inflate(data)
  return parseH3M(content.buffer)
}

export function parseAsset<T>(type: TagOrWrapper<T>, data: StreamInput): T {
  return parse(type, data)
}
