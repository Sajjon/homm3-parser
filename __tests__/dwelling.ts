import { testMap } from './utils'
import { H3MFile } from '../src'

describe('dwelling', () => {
  testMap('dwelling', 'dwelling.h3m', (map: H3MFile) => map.objectDetails)
})
