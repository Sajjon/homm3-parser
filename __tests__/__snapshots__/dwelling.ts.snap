// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dwelling should parse dwelling 1`] = `
Object {
  "count": 27,
  "entries": Array [
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen0.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwelling",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "RandomDwelling",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Castle",
            ],
            "castleAbsodId": 0,
          },
          "level": Object {
            "max": 6,
            "min": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwelling",
      },
      "oaIndex": 2,
      "x": 2,
      "y": 11,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen0.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwelling",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "RandomDwelling",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Rampart",
            ],
            "castleAbsodId": 0,
          },
          "level": Object {
            "max": 6,
            "min": 1,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwelling",
      },
      "oaIndex": 2,
      "x": 5,
      "y": 11,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen0.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwelling",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "RandomDwelling",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Tower",
            ],
            "castleAbsodId": 0,
          },
          "level": Object {
            "max": 6,
            "min": 2,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwelling",
      },
      "oaIndex": 2,
      "x": 8,
      "y": 11,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen0.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwelling",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "RandomDwelling",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Inferno",
            ],
            "castleAbsodId": 0,
          },
          "level": Object {
            "max": 6,
            "min": 3,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwelling",
      },
      "oaIndex": 2,
      "x": 11,
      "y": 11,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen0.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwelling",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "RandomDwelling",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Necropolis",
            ],
            "castleAbsodId": 0,
          },
          "level": Object {
            "max": 6,
            "min": 4,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwelling",
      },
      "oaIndex": 2,
      "x": 14,
      "y": 11,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen0.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwelling",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "RandomDwelling",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Dungeon",
            ],
            "castleAbsodId": 0,
          },
          "level": Object {
            "max": 6,
            "min": 5,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwelling",
      },
      "oaIndex": 2,
      "x": 17,
      "y": 11,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen0.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwelling",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "RandomDwelling",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Stronghold",
            ],
            "castleAbsodId": 0,
          },
          "level": Object {
            "max": 6,
            "min": 6,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwelling",
      },
      "oaIndex": 2,
      "x": 20,
      "y": 11,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen0.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwelling",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "RandomDwelling",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Fortress",
              "Conflux",
            ],
            "castleAbsodId": 0,
          },
          "level": Object {
            "max": 1,
            "min": 1,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwelling",
      },
      "oaIndex": 2,
      "x": 23,
      "y": 11,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen0.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwelling",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "RandomDwelling",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Conflux",
            ],
            "castleAbsodId": 0,
          },
          "level": Object {
            "max": 2,
            "min": 1,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwelling",
      },
      "oaIndex": 2,
      "x": 26,
      "y": 11,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen1.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingLevel",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "RandomDwellingLevel",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Castle",
              "Rampart",
              "Tower",
              "Inferno",
              "Necropolis",
              "Dungeon",
              "Stronghold",
              "Fortress",
              "Conflux",
            ],
            "castleAbsodId": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingLevel",
      },
      "oaIndex": 3,
      "x": 2,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen2.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingLevel",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "RandomDwellingLevel",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Castle",
              "Rampart",
              "Tower",
              "Inferno",
              "Necropolis",
              "Dungeon",
              "Stronghold",
              "Fortress",
              "Conflux",
            ],
            "castleAbsodId": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingLevel",
      },
      "oaIndex": 4,
      "x": 5,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen3.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingLevel",
        "objectGroup": "",
        "objectSubType": 2,
        "objectType": "RandomDwellingLevel",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Castle",
              "Rampart",
              "Tower",
              "Inferno",
              "Necropolis",
              "Dungeon",
              "Stronghold",
              "Fortress",
              "Conflux",
            ],
            "castleAbsodId": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingLevel",
      },
      "oaIndex": 5,
      "x": 8,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen4.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingLevel",
        "objectGroup": "",
        "objectSubType": 3,
        "objectType": "RandomDwellingLevel",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Castle",
              "Rampart",
              "Tower",
              "Inferno",
              "Necropolis",
              "Dungeon",
              "Stronghold",
              "Fortress",
              "Conflux",
            ],
            "castleAbsodId": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingLevel",
      },
      "oaIndex": 6,
      "x": 11,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen5.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingLevel",
        "objectGroup": "",
        "objectSubType": 4,
        "objectType": "RandomDwellingLevel",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Castle",
              "Rampart",
              "Tower",
              "Inferno",
              "Necropolis",
              "Dungeon",
              "Stronghold",
              "Fortress",
              "Conflux",
            ],
            "castleAbsodId": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingLevel",
      },
      "oaIndex": 7,
      "x": 14,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen6.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingLevel",
        "objectGroup": "",
        "objectSubType": 5,
        "objectType": "RandomDwellingLevel",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Castle",
              "Rampart",
              "Tower",
              "Inferno",
              "Necropolis",
              "Dungeon",
              "Stronghold",
              "Fortress",
              "Conflux",
            ],
            "castleAbsodId": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingLevel",
      },
      "oaIndex": 8,
      "x": 17,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen7.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingLevel",
        "objectGroup": "",
        "objectSubType": 6,
        "objectType": "RandomDwellingLevel",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": Array [
              "Castle",
              "Rampart",
              "Tower",
              "Inferno",
              "Necropolis",
              "Dungeon",
              "Stronghold",
              "Fortress",
              "Conflux",
            ],
            "castleAbsodId": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingLevel",
      },
      "oaIndex": 9,
      "x": 20,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgn00.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingAlignment",
        "objectGroup": "",
        "objectSubType": "Castle",
        "objectType": "RandomDwellingAlignment",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "level": Object {
            "max": 6,
            "min": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingAlignment",
      },
      "oaIndex": 10,
      "x": 2,
      "y": 22,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgn01.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingAlignment",
        "objectGroup": "",
        "objectSubType": "Rampart",
        "objectType": "RandomDwellingAlignment",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "level": Object {
            "max": 6,
            "min": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingAlignment",
      },
      "oaIndex": 11,
      "x": 5,
      "y": 22,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgn02.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingAlignment",
        "objectGroup": "",
        "objectSubType": "Tower",
        "objectType": "RandomDwellingAlignment",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "level": Object {
            "max": 6,
            "min": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingAlignment",
      },
      "oaIndex": 12,
      "x": 8,
      "y": 22,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgn03.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingAlignment",
        "objectGroup": "",
        "objectSubType": "Inferno",
        "objectType": "RandomDwellingAlignment",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "level": Object {
            "max": 6,
            "min": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingAlignment",
      },
      "oaIndex": 13,
      "x": 11,
      "y": 22,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgn04.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingAlignment",
        "objectGroup": "",
        "objectSubType": "Necropolis",
        "objectType": "RandomDwellingAlignment",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "level": Object {
            "max": 6,
            "min": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingAlignment",
      },
      "oaIndex": 14,
      "x": 14,
      "y": 22,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgn05.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingAlignment",
        "objectGroup": "",
        "objectSubType": "Dungeon",
        "objectType": "RandomDwellingAlignment",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "level": Object {
            "max": 6,
            "min": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingAlignment",
      },
      "oaIndex": 15,
      "x": 17,
      "y": 22,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgn06.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingAlignment",
        "objectGroup": "",
        "objectSubType": "Stronghold",
        "objectType": "RandomDwellingAlignment",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "level": Object {
            "max": 6,
            "min": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingAlignment",
      },
      "oaIndex": 16,
      "x": 20,
      "y": 22,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgn07.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingAlignment",
        "objectGroup": "",
        "objectSubType": "Fortress",
        "objectType": "RandomDwellingAlignment",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "level": Object {
            "max": 6,
            "min": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingAlignment",
      },
      "oaIndex": 17,
      "x": 23,
      "y": 22,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgn08.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwellingAlignment",
        "objectGroup": "",
        "objectSubType": "Conflux",
        "objectType": "RandomDwellingAlignment",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "level": Object {
            "max": 6,
            "min": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwellingAlignment",
      },
      "oaIndex": 18,
      "x": 26,
      "y": 22,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          192,
        ],
        "allowedLandscapes": 511,
        "header": "avrcgen0.def",
        "landscapeGroup": 255,
        "metaType": "RandomDwelling",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "RandomDwelling",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          63,
        ],
      },
      "body": Object {
        "data": Object {
          "castle": Object {
            "alignments": null,
            "castleAbsodId": 2282018828,
          },
          "level": Object {
            "max": 6,
            "min": 0,
          },
          "owner": "Neutral",
        },
        "type": "RandomDwelling",
      },
      "oaIndex": 2,
      "x": 2,
      "y": 28,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          32,
        ],
        "allowedLandscapes": 255,
        "header": "AVCranx0.def",
        "landscapeGroup": 255,
        "metaType": "Town",
        "objectGroup": "Towns",
        "objectSubType": 0,
        "objectType": "RandomTown",
        "passable": Array [
          255,
          255,
          255,
          143,
          7,
          7,
        ],
      },
      "body": Object {
        "data": Object {
          "absodId": 2282018828,
          "alignment": 255,
          "buildings": null,
          "buildingsCustomized": false,
          "eventCount": 0,
          "events": Array [],
          "formation": 0,
          "garrison": null,
          "hasFort": true,
          "hasGarrison": 0,
          "hasName": 0,
          "mayHaveSpells": Array [],
          "mustHaveSpells": Array [],
          "name": null,
          "owner": "Neutral",
        },
        "type": "Town",
      },
      "oaIndex": 19,
      "x": 9,
      "y": 32,
      "z": 0,
    },
  ],
}
`;
